Optimized graph algo

jdbancal opened the pull request:
Fixes #158

jdbancal commented:
Class `replab.Partitions` has been revamped, please check if this is compatible with your original intent @denisrosset

denisrosset commented:
Thanks for this contribution! Note that I had the idea of storing the partition blocks implicitly; as you added the `blocks` property, I wonder if we may discard the `next` property and simply store the blocks.

jdbancal commented:
Concerning the question of whether having a `blocks` property or a `blocks` function, the main reason I implemented the property here is because this corresponds to the information that we start with, and thus it seems more direct to keep it and use it when we need it rather than erasing this information and later recompute `blocks` on demand. This also does not require additional memory since storing the blocks takes basically the same memory as the `first` and `next` vectors that we already have. So I see no loss here, only a gain in processing effort.

If the main purpose of the `next` property was to recompute the `blocks`, then I agree we can remove it. I'm not sure if the `first` property was also introduced for the implicit definition of blocks. Do you want to keep the `first` property? If needed, it can be replaced by something like `cellfun(@(x) x(1), blocks)`. I'm also fine with keeping it since it has a low footprint.

So far, I did not experience a bottleneck with the active block recomputation, but I also haven't used it really. To me, the question is rather whether we want to design our workflow with a computation step that we don't really need since we already knew the answer to the question (and keeping it takes up no additional memory)...

If I understand, you mention that the Matlab just-in-time compiler optimizes poorly code involving cell arrays of vectors. In which case did you see something like this happen? Do you have a concrete example? To iterate on all elements of a `Partition` block by block, for instance, I could expect a cell-array-based memory structure to be more efficient than a next-vector based one, because successive elements are most of the time located next to each other in memory, but I haven't checked.

Yes, of course we can start to discuss about approaches to optimization strategies on skype! Personally, I am not aware if there are some parts of the code that you would be thinking of implementing in java, so we should definitely talk about that! Since the current code is a first instance of hard optimization in RepLAB, I'ld also be fine with incorporating it in develop while making clear that its final form will be subject to the result of the discussion we are starting: in any case, some sort of optimization will be necessary for the task at hand here, and this can be modified later if we find a better way of doing it.
